src/internal/storage/chunk/client.go

```diff
@@ -132,6 +166,54 @@ func (c *trackedClient) Close() error {
 	return nil
 }
 
+// Check checks that there are entries for the chunk and they have successfully uploaded objects.
+func (c *trackedClient) Check(ctx context.Context, id ID, readChunk bool) error {
+	last, err := c.CheckEntries(ctx, id, 1, readChunk)
+	if err != nil {
+		return err
+	}
+	if last == nil {
+		return errors.Errorf("no entries for chunk %v", id)
+	}
+	return nil
+}
+
+// CheckEntries runs an integrity check on the objects in object storage.
+// It lists through chunks with IDs >= first, lexicographically.
+func (c *trackedClient) CheckEntries(ctx context.Context, first []byte, limit int, readChunks bool) (last ID, _ error) {
```

Comment:
Wouldn't this produce an error if a chunk was correctly deleted between the query and the check? It seems like it should requery after an existence error to ensure that it wasn't deleted from the chunk table. Also, why is the iteration exposed through multiple layers like this (the first / last stuff). Is the concern about a long running transaction? If so, I'm of the opinion that it is fine to have a few things in the system that have longer running transactions (such as GC) in the interest of simplicity. If the concern is performance and the results becoming stale, this might be another use case for the task service similar to GC.

Reply:
Lol, just realized the extra check is the TODO in the comment. My bad. The iteration strategy question still stands.

Reply:
I wanted to capture the interactions with the chunk table in one place. Since these entries can be in flux it doesn't really make sense to hold open a connection.  It's better to do a few rows at a time, then get a few more.   We can make the iteration at the next layer up more complicated, or add parallelism.